### DISCUSSION TITLE
Wallet v2 features and UX

### DISCUSSION THREAD

**Discussion by elsirion** [fedimint/fedimint#6413] [General]
There's a previous discussion topic #5893, but the outlined design changed considerably in the meantime as @bradleystachurski and @joschisan kept iterating on it. The topic of Wallet v2 also has become more actionable in the meantime and @joschisan already has a prototype (#6429). So I'd like to open a discussion about the trade-offs involved:

# New protocol
Major proposed changes:
* Federation only keeps one UTXO 
  * Deposits are merged into it instantly
  * Withdraws will be built with a potentially unconfirmed input UTXO, meaning there are on-chain fees on deposit now
  * This means coin selection becomes trivial and fees can be bumped through CPFP
* No txoutproofs anymore, federation keeps track of partial UTXO set (same as done in #6382 for Wallet v1)

That protocol would be massively simpler than the old one, to be laid out here (@joschisan @bradleystachurski ): …

## Deposits
* Users still generate addresses by tweaking the federation descriptor
* After depositing, users still wait for `finality_delay + 1` confirmations
* User queries on-chain fee rate to create the merge transaction
* User sends Fedimint claim transaction, revealing the tweak in return for e-cash (minus on-chain fees for merging the UTXO into the federation UTXO)
* User receives e-cash

## Withdraws
* User queries on-chain fee rate
  * If the federation UTXO is unconfirmed the user will have to also pay for the unconfirmed parent transactions to get the total package fee rate high enough to be confirmed in a timely manner
* User funds a fedimint withdraw transaction with e-cash and specifies destination address (funding=fees+withdraw amount)
* The federation constructs a transaction with its one UTXO as input (may be unconfirmed) and the user destination and the new federation UTXO as outputs

# UX considerations

Unfortunately, it's not possible to inform about deposit fees upfront, so it's hard to automate completely. I see two main options with a bunch of variations thereof in between:

## Option 1: User has full control:
* Deposit address is generated
* User deposits
* User can view all generated addresses
* User can click any address to see received transactions
* Once a transaction is confirmed, the user has to find it that way and claim the deposit manually
* They'll be told the deposit fee at that point since it's not predictable before starting the deposit
* They can choose to wait for a lower fee environment
* Benefit: simpler on the Fedimint side
* Drawback: more convoluted UX

## Option 2: Automated
* Deposit address is generated
* User deposits, we can't tell the fee they will pay, but maybe let them set an upper bound
* Deposit is automatically claimed once the transaction confirms as long as fees are low enough at the time
* If fees are higher than the upper bound we still need to let the user make a choice if they want to claim or wait for lower fees
* Drawbacks: 
  * We lose some privacy if multiple deposits are in flight simultaneously
  * Still needs fallback UX if a) a deposit address times out or b) the fees are too high to automatically claim

I think Option 2 can be built from Option 1 quite easily, just if every integrator end up building their own automation I'd rather have it in Fedimint.

# Other considerations
*To be amended as feedback comes in*

* Feedback from one guardian: not comfortable with merging deposits into one UTXO, could taint federation coins. They'd rather see us giving more control over UTXO management to guardians.
* The question of backwards compatibility and smooth upgrade paths came up inside Fedi, the two options I see:
  * Find a way to make wallet v2 really more of a minor consensus version upgrade of v1. We'd first add client compatibility and later make guardians automatically upgrade their wallet consensus version.
  * Keep supporting wallet v1 and v2 in parallel indefinitely (until the last v1 federation shuts down).

**Comment by joschisan** [comment]
The current implementation derives the following number of vbytes for send an receive transaction:

| Guardians | Send | Receive |
|-----------|------|---------|
| 1         | 166  | 192     |
| 4         | 228  | 316     |
| 5         | 255  | 369     |
| 6         | 281  | 423     |
| 7         | 290  | 440     |
| 8         | 317  | 494     |
| 9         | 344  | 548     |
| 10        | 352  | 565     |
| 11        | 379  | 618     |
| 12        | 406  | 672     |
| 13        | 414  | 689     |
| 14        | 441  | 742     |
| 15        | 468  | 796     |
| 16        | 476  | 813     |
| 17        | 503  | 867     |
| 18        | 530  | 920     |
| 19        | 539  | 937     |
| 20        | 565  | 991     |

Due to the witness discount the vbytes per cosigner / guardian actually decrease with every additional guardian. Furthermore, we use an adaptive multiplier now for the feerate instead of the static 4x. Currently, we multiply the ferrate by 4/3 for every currently unconfirmed transaction. This results in the following multipliers with respect to the height of the stack of unconfirmed transactions: 

Unconfirmed | Last Multiplier | Average Multiplier
------------|-----------------|-------------------
1           | 1.00            | 1.00
2           | 1.33            | 1.17
3           | 1.78            | 1.37
4           | 2.37            | 1.62
5           | 3.16            | 1.93
6           | 4.21            | 2.31
7           | 5.62            | 2.78
8           | 7.49            | 3.28

Therefore the walletv2 fees should actually be quite reasonable.

**Comment by joschisan** [comment]
> I think Option 2 can be built from Option 1 quite easily, just if every integrator end up building their own automation I'd rather have it in Fedimint.

If you build an automation you need to make a few non-obvious choices regarding fees and privacy. Are you setting an absolute or relative limit for automatic receive, or both? How often do you check a given address given its time of generation, index and prior use?

If the integrators build the automation it requires less communication and makes it more obvious what the fedimint client is actually doing.

**Comment by dpc** [comment]
Idea:

Since federation already needs to store the list of all the confirmed potential peg-in txes, we could entirely eliminate client needing to ever query external Bitcoin blockchain datasources, and just query Federation for everything.

All that is required is one extra index, reverse to what is already handled by the walletv2 impl. Very easy change, but double the storage cost (non-issue, IMO).

This would not let the client query for pending (in mempool) txes, which is meh UX. To help with that, federation members could just periodically query and keep around a list of txes in the mempool that could potentially be peg-ins. This logic would be entirely non-consensus and non-correctness-critical. All it would do is the help with UX, so the user can see their peg-ins in the mempool faster.

If I'm not mistaken this would greatly improve privacy and stop clients relying on Esplora (by default) for everything.

@joschisan was initially against it, but I think it's a great client side improvement at minimum server side cost.

@elsirion @bradleystachurski for awarness.

**Reply by elsirion** [reply]
Where's the privacy improvement? Guardians could already run their own esplora server for users if they are more trusted. I think in the end you leak the same data, just to different entities.

**Reply by dpc** [reply]
@elsirion 

* We have Iroh/Tor connectivity to guardians built-in.
* Queries for on-chain stuff from the outside look exactly like any other queries.
* We don't involve any extra parties.
* More reliable - if the Federation is up, Esplora being down can't mess with anything. Threshold redundancy still applies too.
* We don't default to have *all* clients query the same centralized entity.
